src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/Pipeline.cs

```diff
@@ -0,0 +1,547 @@
+            /// <since_tizen> 8 </since_tizen>
+            public void Control(bool open)
+            {
+                if (!Valid)
+                    throw NNStreamerExceptionFactory.CreateException(NNStreamerError.InvalidOperation, "Current node is invalid");
+
+                NNStreamerError ret = Interop.Pipeline.OpenValve(Handle, open);
+                NNStreamer.CheckException(ret, "Failed to set valve status: " + Name);
+            }
+        }
+
+        /// <summary>
+        /// SourceNode class to handle the source node.
+        /// </summary>
+        /// <since_tizen> 8 </since_tizen>
+        public class SourceNode : NodeInfo
+        {
+            internal SourceNode(string name, Pipeline pipe) : base(NodeType.Source, name, pipe)
+            {
+                IntPtr handle = IntPtr.Zero;
+
+                NNStreamerError ret = Interop.Pipeline.GetSrcHandle(pipe.GetHandle(), name, out handle);
+                NNStreamer.CheckException(ret, "Failed to get the source node handle: " + name);
+
+                Handle = handle;
+            }
+
+            /// <summary>
+            /// Inputs tensor data to source node.
+            /// </summary>
+            /// <param name="data">The tensors data</param>
+            /// <feature>http://tizen.org/feature/machine_learning.inference</feature>
+            /// <exception cref="NotSupportedException">Thrown when the feature is not supported.</exception>
+            /// <exception cref="ArgumentException">Thrown when the method failed due to an invalid parameter.</exception>
+            /// <exception cref="InvalidOperationException">Thrown when the node is invalid, or failed to input tensors data.</exception>
+            /// <since_tizen> 8 </since_tizen>
+            public void Input(TensorsData data)
+            {
+                if (data == null)
+                    throw NNStreamerExceptionFactory.CreateException(NNStreamerError.InvalidParameter, "Given data is invalid");
+
+                if (!Valid)
+                    throw NNStreamerExceptionFactory.CreateException(NNStreamerError.InvalidOperation, "Current node is invalid");
+
+                data.PrepareInvoke();
+
+                NNStreamerError ret = Interop.Pipeline.InputSrcData(Handle, data.GetHandle(), PipelineBufferPolicy.NotFreed);
+                NNStreamer.CheckException(ret, "Failed to input tensors data to source node: " + Name);
+            }
+        }
+
+        /// <summary>
+        /// SinkNode class to handle the sink node.
+        /// </summary>
+        /// <since_tizen> 8 </since_tizen>
+        public class SinkNode : NodeInfo
+        {
+            private event EventHandler<DataReceivedEventArgs> _dataReceived;
+            private Interop.Pipeline.NewDataCallback _dataCallback;
+
+            internal SinkNode(string name, Pipeline pipe) : base(NodeType.Sink, name, pipe)
```

Comment:
SinkNode does not check validation of node name when it was created.
So, exception was not thrown when Sink node was created
``` c#
var sink = pipeline.GetSink("invalid name"); // no exception

sink.DataReceived += (s, e) => { .... }; // exception will be thrown
```
app developer hard to know that sinknode name is wrong, just feels DataReceived event not working.

Reply:
You are right. I just updated PR to get the sink node handle from native.